File: tree/practice/longest-increasing-subsequence/longest-increasing-subsequence.java
class Solution {
    private int bsToFindSmallestGreaterThanEqNum(List<Integer> subseq, int num){
        int left = 0;
        int right = subseq.size();
        while(left < right){
            int mid = left + (right-left)/2;
            if(subseq.get(mid) == num) return mid;
            else if(subseq.get(mid) < num) left = mid+1;
            else right = mid;
        }
        return left;
    }
    public int lengthOfLIS(int[] nums) {
        int len = nums.length;
        if(len == 0) return 0;

        List<Integer> subseq = new ArrayList<Integer>();
        subseq.add(nums[0]);

        for(int i = 1; i < len; i++){
            if(nums[i] > subseq.get(subseq.size()-1)){
                subseq.add(nums[i]);
            } else {
                int indexToReplace = bsToFindSmallestGreaterThanEqNum(subseq, nums[i]);
                subseq.set(indexToReplace, nums[i]);
            }
        }
        return subseq.size();
    }
}

File: tree/practice/longest-increasing-subsequence/longest-increasing-subsequence.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        def binary_search(subseq, target):
            l, r = 0, len(subseq) - 1
            while l < r:
                m = (l+r)//2
                if subseq[m] < target:
                    l = m + 1
                else:
                    r = m
            return l
        
        track = []
        for n in nums:
            if not track or n > track[-1]:
                track.append(n)
                continue
            idx = binary_search(track, n)
            track[idx] = n
        return len(track)